src/cluster/placement/algo/mirrored_test.go

```diff
@@ -1619,6 +1624,121 @@ func TestMarkInstanceAndItsPeersAvailable(t *testing.T) {
 	})
 }
 
+func TestBalanceShardsForMirroredWhenBalanced(t *testing.T) {
+	i1 := newTestInstance("i1").
+		SetShardSetID(1).
+		SetWeight(1).
+		SetShards(shard.NewShards([]shard.Shard{
+			shard.NewShard(0).SetState(shard.Available),
+		}))
+	i2 := newTestInstance("i2").
+		SetShardSetID(1).
+		SetWeight(1).
+		SetShards(shard.NewShards([]shard.Shard{
+			shard.NewShard(0).SetState(shard.Available),
+		}))
+	i3 := newTestInstance("i3").
+		SetShardSetID(2).
+		SetWeight(1).
+		SetShards(shard.NewShards([]shard.Shard{
+			shard.NewShard(1).SetState(shard.Available),
+		}))
+	i4 := newTestInstance("i4").
+		SetShardSetID(2).
+		SetWeight(1).
+		SetShards(shard.NewShards([]shard.Shard{
+			shard.NewShard(1).SetState(shard.Available),
+		}))
+	initialPlacement := placement.NewPlacement().
+		SetReplicaFactor(2).
+		SetShards([]uint32{0, 1}).
+		SetInstances([]placement.Instance{i1, i2, i3, i4}).
+		SetIsMirrored(true).
+		SetIsSharded(true).
+		SetMaxShardSetID(2)
+
+	expectedPlacement := initialPlacement.Clone()
+
+	a := NewAlgorithm(placement.NewOptions().SetIsMirrored(true))
+
+	balancedPlacement, err := a.BalanceShards(initialPlacement)
+	assert.NoError(t, err)
+	assert.Equal(t, expectedPlacement, balancedPlacement)
+}
+
+func TestBalanceShardsForMirroredWhenImbalanced(t *testing.T) {
```

Comment:
Not sure how we could end up with imbalanced placement in the mirrored algo because removal/reclaiming is blocking until all affected shards are cutover. Could you add a test that starts with a balanced placement and simulates a sequence of operations that results in imbalanced placement?

Reply:
I don't quite see the part in the mirrored algo which would block add/remove operations until all shards are cutover. Where is that happening?

I'm not if it would be worth having such a test case because it would be rather difficult to produce (as we would need add/remove operations to take some time) and it would not really test much.

I agree that there should not be any shard imbalance if add/remove operations are performed one after another. However, that could happen if such an operation is manually interrupted in the middle and then reverted (by utilizing "reclaimLeaving" functionality). We have encountered such scenarios for clusters relying the sharded algo and I believe it could happen for the mirrored algo as well because it also uses sharded algo underneath.

Reply:
As discussed offline, neither for sharded, nor mirrored algorithms shard imbalance could occur w/o human operator involvement. However, in both cases shard imbalance could occur if an existing instance for which a topology change (e.g. replacement) is being performed would be re-added and then its leaving shards would be reclaimed.

As you pointed out, certain simultaneous topology changes e.g. another replacement while an add is ongoing would be blocked, however, that does not apply here. That's because in the scenario that might cause shard imbalance there would be no real topology changes but rather just "reclaimLeavingShards" call triggered by a call to `AddInstances` or `ReplaceInstances` with instances that already exist in placement (and have shards moving). Such calls might be potentially useful to revert an ongoing topology change (e.g. if one was started incorrectly) and there is no validation blocking them.